ceramic-kubo-rpc/src/http/block.rs

```diff
@@ -0,0 +1,291 @@
+use std::str::FromStr;
+
+use actix_multipart::Multipart;
+use actix_web::{http::header::ContentType, web, HttpResponse, Scope};
+use anyhow::anyhow;
+use dag_jose::DagJoseCodec;
+use futures_util::StreamExt;
+use iroh_api::Cid;
+use libipld::{cbor::DagCborCodec, json::DagJsonCodec};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    block,
+    error::Error,
+    http::{AppState, DAG_CBOR, DAG_JOSE, DAG_JSON},
+    IpfsDep,
+};
+pub fn scope<T>() -> Scope
+where
+    T: IpfsDep + 'static,
+{
+    web::scope("/block")
+        .service(web::resource("/get").route(web::post().to(get::<T>)))
+        .service(web::resource("/put").route(web::post().to(put::<T>)))
+        .service(web::resource("/stat").route(web::post().to(stat::<T>)))
+}
+
+const RAW: &str = "raw";
+const SHA2_256: &str = "sha2-256";
+
+// used to provide default to query structs
+fn raw() -> String {
+    RAW.to_string()
+}
+fn sha2_256() -> String {
+    SHA2_256.to_string()
+}
+
+#[derive(Debug, Deserialize)]
+struct GetQuery {
+    arg: String,
+}
+
+#[tracing::instrument(skip(data))]
+async fn get<T>(
+    data: web::Data<AppState<T>>,
+    query: web::Query<GetQuery>,
+) -> Result<HttpResponse, Error>
+where
+    T: IpfsDep,
+{
+    let cid = Cid::from_str(query.arg.as_str()).map_err(|e| Error::Invalid(e.into()))?;
+
+    Ok(HttpResponse::Ok()
+        // Note the content is not really text/plain, however the Kubo RPC
+        // explicitly specifies the text/plain content type.
+        .content_type(ContentType(mime::TEXT_PLAIN))
+        .body(block::get(data.api.clone(), cid).await?))
+}
+
+#[derive(Debug, Deserialize)]
+struct PutQuery {
+    #[serde(rename = "cid-codec", default = "raw")]
+    cid_codec: String,
+    #[serde(default = "sha2_256")]
+    mhtype: String,
+    #[serde(default)]
+    pin: bool,
+    format: Option<String>,
+}
+
+#[tracing::instrument(skip(data, payload))]
+async fn put<T>(
+    data: web::Data<AppState<T>>,
+    query: web::Query<PutQuery>,
+    mut payload: Multipart,
+) -> Result<HttpResponse, Error>
+where
+    T: IpfsDep,
+{
+    if query.pin {
+        return Err(Error::Invalid(anyhow!(
+            "recursive pinning is not supported."
```

Comment:
Does `pin` being `true` here always mean recursively?

Reply:
Yes, according to their docs https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-block-put We might need to implement recursive pin but I am hoping not too. So much of our data points back to the beginning of time that its dangerous to recursively pin so for now I have explicitly denied any recursive pin behavior.